api/src/main/java/org/openmrs/module/appframework/service/AppFrameworkServiceImpl.java

```diff
@@ -356,6 +369,62 @@ public boolean checkRequireExpression(Extension candidate, AppContextModel conte
         }
     }
 
+    protected String getPatientUuid(AppContextModel contextModel) {
+    	Bindings bindings = new SimpleBindings();
+	    for (Map.Entry<String, Object> e : contextModel.entrySet()) {
+             bindings.put(e.getKey(), e.getValue());
+	    }
+        javascriptEngine.setBindings(bindings, ScriptContext.ENGINE_SCOPE);
+    	try {
+			String uuid = (String) javascriptEngine.eval("patient.uuid;");
+			if (StringUtils.isBlank(uuid)) {
+	    		throw new APIException("Patient uuid cannot be empty");
+			}
+			return uuid;
+		} catch (ScriptException e) {
+    		    	throw new APIException("Failed to get patient uuid", e);
+		}
+    }
+    
+    protected boolean checkRequiredState(Extension candidate, AppContextModel contextModel) {
+    	String requiredState = candidate.getRequiredState();
+    	if (StringUtils.isBlank(requiredState)) {
+    		return true;
+    	}
+    	Concept stateConcept = AppFrameworkUtil.getConcept(requiredState);
+    	if (stateConcept == null) {
+    		throw new APIException("Could not find Concept with id or mapping :" + requiredState);
+    	}
+    	String patientUuid = getPatientUuid(contextModel);
+ 
+    	Patient patient = Context.getPatientService().getPatientByUuid(patientUuid);
+    	if (patient == null) {
+    		throw new APIException("Could not find patient with uuid :" + patientUuid);
+    	}
+    	List<PatientProgram> patientPrograms = Context.getProgramWorkflowService().getPatientPrograms(patient, null, null,
+			    null, null, null, false);
+    	Set<PatientState> states = new HashSet<PatientState>();
+    	for (PatientProgram patientProgram : patientPrograms) {
+    		Program program = patientProgram.getProgram();
+    		Set<ProgramWorkflow> workflows = program.getAllWorkflows();
+    		if (workflows != null) {
+    			for (ProgramWorkflow workflow : workflows) {
+    				PatientState state = patientProgram.getCurrentState(workflow);
+    				if (state != null) {
+    					states.add(state);
```

Comment:
I would check here that `state`'s `ProgramWorkflowState` matches the one obtained earlier. If it matches, you're good and you can get out early: the required state condition is met. No?

Reply:
Correct!, nice catch.